impl/monitor/monitor.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package monitor implements the monitor service. A monitor repeatedly polls a
+// key-transparency server's Mutations API and signs Map Roots if it could
+// reconstruct
+// clients can query.
+package monitor
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/google/trillian"
+
+	ktpb "github.com/google/keytransparency/core/proto/keytransparency_v1_types"
+	tv "github.com/google/keytransparency/core/tree/sparse/verifier"
+
+	mspb "github.com/google/keytransparency/impl/proto/monitor_v1_service"
+	mupb "github.com/google/keytransparency/impl/proto/mutation_v1_service"
+
+	"bytes"
+	"github.com/google/keytransparency/core/tree/sparse"
+)
+
+// Each page contains pageSize profiles. Each profile contains multiple
+// keys. Assuming 2 keys per profile (each of size 2048-bit), a page of
+// size 16 will contain about 8KB of data.
+const pageSize = 16
+
+// Server holds internal state for the monitor server.
+type server struct {
+	client     mupb.MutationServiceClient
+	pollPeriod time.Duration
+
+	tree *tv.Verifier
+	// TODO(ismail) abstract this into a storage interface and have an in-memory version
+	seenSMRs          []*trillian.SignedMapRoot
```

Comment:
Maybe create a struct for this:
- SMR
- Was it valid? / Error type / status
- Time observed